### PR TITLE
Using a new user per feature, to facilitate parallelization of secondary_parallelizable.yml in CI

### PR DESCRIPTION
**Pull request by srbarrios**
## What does this PR change?

Related to https://github.com/SUSE/spacewalk/issues/9748

Creating a new user per feature, using a scenario hook that will only pass for the first scenario per feature, acting as a feature hook.

It will by-pass the authentication of testing or admin users, using the regular step, when the feature variable @username is not null.

It creates a user name based on the feature filename because the feature name is too long and we have secondary features that acts on several clients during the same feature, so client name can't be use for that purpose on regular CI Tests. 

How to debug parallel tests is another topic, that we can discuss and improve separetely.
The aime of that PR is only to fix the current issue with multiple systems selected as described in the issue linked.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed

- [x] **DONE**

## Test coverage
- Cucumber tests were changed, need to check on CI

- [x] **DONE**

## Links

None

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"		 
- [ ] Re-run test "java_pgsql_tests"		 
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"		 
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"		 
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by srbarrios** [comment]
https://ci.suse.de/blue/organizations/jenkins/uyuni-prs-ci-tests/detail/uyuni-prs-ci-tests/1513/pipeline

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 1]
```diff
@@ -396,7 +396,13 @@
 Given(/^I am authorized for the "([^"]*)" section$/) do |section|
   case section
   when 'Admin'
-    step %(I am authorized as "admin" with password "admin")
+    if $username
+      step %(I am authorized as "#{$username}" with password "#{$password}")
+      step %(I am logged in API as user "#{$username}" and password "#{$password}")
```
Comment (Bischoff) [inline testsuite/features/step_definitions/navigation_steps.rb:401]: isn't that redundant?

**Review by Bischoff — COMMENTED** [review]
see question, other than that lgtm

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 1 above)
Comment (srbarrios) [reply testsuite/features/step_definitions/navigation_steps.rb:401]: Not really, one is log-in from the WebUI so in the browser.
The other is log-in from the API, so keeping a connection established in our api_client

File: testsuite/features/step_definitions/salt_steps.rb [hunk 2]
```diff
@@ -162,12 +162,23 @@
...
+  $api_test.user.add_role($username, 'org_admin')
+  $api_test.user.add_role($username, 'channel_admin')
+  $api_test.user.add_role($username, 'config_admin')
+  $api_test.user.add_role($username, 'system_group_admin')
+  $api_test.user.add_role($username, 'activation_key_admin')
+  $api_test.user.add_role($username, 'image_admin')
+  puts "New user #{$username} created"
+end
+
+Then(/^I can cleanup the no longer needed user$/) do
```
Comment (nodeg) [inline testsuite/features/step_definitions/salt_steps.rb:180]: I did not see this step used anywhere.

Comment (srbarrios) [reply testsuite/features/step_definitions/salt_steps.rb:180]: Uhm, it seems a leftover from the old master I had to fix conflicts. I will remove this step.

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 1 above)
Comment (Bischoff) [reply testsuite/features/step_definitions/navigation_steps.rb:401]: ah yes of course

but then i would do the same in case 'Images' just below, for consistency

Comment (srbarrios) [reply testsuite/features/step_definitions/navigation_steps.rb:401]: I think kiwikiwi has some special permissions different than admin, also we are using it in very specific test cases.
So for now, better if we don't touch it...

**Review by nodeg — APPROVED** [review]

**Review by Bischoff — CHANGES_REQUESTED** [review]

**Review by nodeg — CHANGES_REQUESTED** [review]
~~I found some issues in the PR tests:~~

It seems this was from an old PR test run!

```bash
[2022-08-05T10:35:01.439Z] Scenario: Add a product and one of its modules                                                       # features/reposync/srv_sync_products.feature:36

[2022-08-05T10:35:01.439Z] This scenario ran at: 2022-08-05 12:34:35 +0200

[2022-08-05T10:35:01.439Z] When I follow the left menu "Admin > Setup Wizard > Products"                                      # features/step_definitions/navigation_steps.rb:342

[2022-08-05T10:35:01.439Z]       Unable to find xpath "//aside/div[@id='nav']/nav/ul/li/div/a[contains(.,'Admin')]/parent::div/parent::li" (Capybara::ElementNotFound)

[2022-08-05T10:35:01.439Z]       ./features/step_definitions/navigation_steps.rb:361:in `block (2 levels) in <top (required)>'

[2022-08-05T10:35:01.439Z]       ./features/step_definitions/navigation_steps.rb:355:in `each'

[2022-08-05T10:35:01.439Z]       ./features/step_definitions/navigation_steps.rb:355:in `each_with_index'

[2022-08-05T10:35:01.439Z]       ./features/step_definitions/navigation_steps.rb:355:in `/^I follow the left menu "([^"]*)"$/'

[2022-08-05T10:35:01.439Z]       features/reposync/srv_sync_products.feature:37:in `I follow the left menu "Admin > Setup Wizard > Products"'
```
~~It could be that this scenario is not done with the admin user and therefore the scenario fails. This is also not the only scenario that failed with this regards.~~

**Review by nodeg — APPROVED** [review]
I approved it but am curious on how this has an impact on the performance. But let`s wait for Erics go/no go.

**Comment by srbarrios** [comment]
> I approved it but am curious on how this has an impact on the performance. But let`s wait for Erics go/no go.

It will for sure increase a bit the time per feature.
But, when we enable parallel tests this will be very compensated :)

**Review by Bischoff — DISMISSED** [review]
I like it

**Comment by srbarrios** [comment]
https://ci.suse.de/blue/organizations/jenkins/uyuni-prs-ci-tests/detail/uyuni-prs-ci-tests/1523/pipeline/

**Comment by srbarrios** [comment]
Testing it here https://ci.suse.de/view/Manager/view/Uyuni-PRs/job/uyuni-prs-ci-tests/2056/

File: testsuite/features/support/env.rb [hunk 3]
```diff
@@ -139,6 +148,22 @@ def enable_assertions
   log "This scenario ran at: #{current_time}\n"
 end
 
+# Create an user for each feature
```
Comment (vandabarata) [inline testsuite/features/support/env.rb:151]:
```suggestion
# Create a user for each feature
```

**Review by vandabarata — DISMISSED** [review]
Great job with this, looking good! Only the tiniest nitpick ever with a small typo on a comment. 
If it passes on the pipeline, it looks ready to be merged :)

File: testsuite/features/support/env.rb (hunk 3 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb:151]: Right! Thank you

**Review by NamelessOne91 — DISMISSED** [review]

**Comment by nodeg** [comment]
This needs a rebase and more testing with the new changes. I converted it to a draft again.

**Review by maximenoel8 — APPROVED** [review]

File: testsuite/features/secondary/min_bootstrap_negative.feature [hunk 4]
```diff
@@ -71,3 +71,29 @@ Feature: Negative tests for bootstrapping normal minions
     Then I should see a "Standard Error" text
     And I should see "port 11: Connection refused" or "port 11: Network is unreachable" in the stderr textarea
     When I close the modal dialog
+
+  Scenario: Cleanup: bootstrap a SLES minion after negative tests
```
Comment (maximenoel8) [inline testsuite/features/secondary/min_bootstrap_negative.feature:75]: Those changes don't seem related to your PR. Is it a leftover or expected ?

File: testsuite/features/secondary/min_salt_mgrcompat_state.feature [hunk 5]
```diff
@@ -106,3 +106,19 @@ Feature: Verify that Salt mgrcompat state works when the new module.run syntax i
     And I click on "Bootstrap"
     And I wait until I see "Bootstrap process initiated." text
     And I wait until onboarding is completed for "sle_minion"
+
+  Scenario: Cleanup: restore channels on the minion after mgrcompat tests
```
Comment (maximenoel8) [inline testsuite/features/secondary/min_salt_mgrcompat_state.feature:110]: Thoses changes don't seem related to you PR.

File: testsuite/features/secondary/min_salt_minions_page.feature [hunk 6]
```diff
@@ -88,3 +88,19 @@ Feature: Management of minion keys
     And I click on "Bootstrap"
     And I wait until I see "Bootstrap process initiated." text
     And I wait until onboarding is completed for "sle_minion"
+
+  Scenario: Cleanup: restore channels on the minion
```
Comment (maximenoel8) [inline testsuite/features/secondary/min_salt_minions_page.feature:92]: Same here.

**Comment by maximenoel8** [comment]
I'm not an expert on the API side of the product so couldn't really comment on the changes.

Also, we will need to remove the login from bv smoke test.

Just one remark, we don't check who is triggering the events. I suppose it should be ok now with the changes on `After do | scenario | ` but by checking which user was triggering the event I catch some testsuite issues ( the testsuite was changing to admin after a fail scenario.). I don't know if it will be complicated to check which user is triggering the event ( like the reboot for sle micro )

**Comment by maximenoel8** [comment]
Also for all the test you are changing, should you not removed this section : 
```
  Scenario: Log in as admin user
    Given I am authorized for the "Admin" section
```

**Comment by srbarrios** [comment]
> Also for all the tests you are changing, you should changed this section no ? :
> 
> ```
>   Scenario: Log in as admin user
>     Given I am authorized for the "Admin" section
> ```
> 
> For example for the BV init clients.

My idea for now is to only change the steps in CI, to don't interfere with the changes made in BV.

**Comment by srbarrios** [comment]
> I'm not an expert on the API side of the product so couldn't really comment on the changes.
> 
> Also, we will need to remove the login from bv smoke test.
> 
> Just one remark, we don't check who is triggering the events. I suppose it should be ok now with the changes on `After do | scenario | ` but by checking which user was triggering the event I catch some testsuite issues ( the testsuite was changing to admin after a fail scenario.). I don't know if it will be complicated to check which user is triggering the event ( like the reboot for sle micro )

Checking the user is doable, and I think it's desired, but as we are passing the user as a text parameter in the .feature file, that's easy only if we use a template for the feature. But we will need to build specific step definitions overloading some of the current steps if we don't use a template. So, I would like to merge this PR without it, to make it simple, and we can iterate over it later.

File: testsuite/features/secondary/min_salt_minions_page.feature (hunk 6 above)
Comment (srbarrios) [reply testsuite/features/secondary/min_salt_minions_page.feature]: Indeed, thank you

File: testsuite/features/secondary/min_bootstrap_negative.feature (hunk 4 above)
Comment (srbarrios) [reply testsuite/features/secondary/min_bootstrap_negative.feature]: Indeed, thank you

File: testsuite/features/secondary/min_salt_mgrcompat_state.feature (hunk 5 above)
Comment (srbarrios) [reply testsuite/features/secondary/min_salt_mgrcompat_state.feature]: Indeed, thank you

File: testsuite/features/support/env.rb [hunk 7]
```diff
@@ -205,6 +220,22 @@ def process_code_coverage
   skip_this_scenario
 end
 
+# Create a user for each feature
+Before do |scenario|
```
Comment (maximenoel8) [inline testsuite/features/support/env.rb:224]: If I'm correct, this step is also going to be executed for secondary non parallel features. You should probably differentiate the secondary and parallel secondary features to not execute for non parallel feature.

File: testsuite/features/support/env.rb [hunk 8]
```diff
@@ -192,7 +201,13 @@ def process_code_coverage
 end
 
 AfterStep do
-  log 'Timeout: Waiting AJAX transition' if has_css?('.senna-loading', wait: 0) && !has_no_css?('.senna-loading', wait: 30)
+  next unless Capybara::Session.instance_created?
```
Comment (maximenoel8) [inline testsuite/features/support/env.rb:204]: This change doesn't look related to the new changes.

File: testsuite/features/support/env.rb [hunk 9]
```diff
@@ -144,9 +150,12 @@ def capybara_register_driver
       attach "#{Time.at(@scenario_start_time).strftime('%H:%M:%S:%L')} - #{Time.at(current_epoch).strftime('%H:%M:%S:%L')} | Current URL: #{current_url}", 'text/plain'
     rescue StandardError => e
       warn e.message
+      register_driver
```
Comment (maximenoel8) [inline testsuite/features/support/env.rb:153]: Those changes don't seem related to the PR.

File: testsuite/features/support/env.rb [hunk 10]
```diff
@@ -98,20 +97,27 @@ def capybara_register_driver
...
+  Capybara.enable_aria_label = true
+  Capybara.automatic_label_click = true
+  Capybara.app_host = "https://#{ENV.fetch('SERVER', nil)}"
+  Capybara.server_port = 8888 + ENV['TEST_ENV_NUMBER'].to_i
+  $stdout.puts "Capybara APP Host: #{Capybara.app_host}:#{Capybara.server_port}"
 end
 
+# register chromedriver headless mode
 $capybara_driver = capybara_register_driver
-Selenium::WebDriver.logger.level = :error unless $debug_mode
```
Comment (maximenoel8) [inline testsuite/features/support/env.rb:106]: Same here

File: testsuite/features/support/env.rb [hunk 11]
```diff
@@ -22,7 +22,6 @@
 # code coverage analysis
 # SimpleCov.start
 
-server = ENV.fetch('SERVER', nil)
```
Comment (maximenoel8) [inline testsuite/features/support/env.rb:25]: Why do you remove this variable ?

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 12]
```diff
@@ -378,6 +378,8 @@
 
 Given(/^I am not authorized$/) do
   begin
+    xpath_logout = '//a[@href=\'/rhn/Logout.do\']'
```
Comment (maximenoel8) [inline testsuite/features/step_definitions/navigation_steps.rb:381]: Why do we have this change ?

**Comment by maximenoel8** [comment]
Overall don't forget to update the copyright.

File: testsuite/features/support/env.rb [hunk 13]
```diff
@@ -98,20 +97,27 @@ def capybara_register_driver
       unhandledPromptBehavior: 'accept'
     )
 
-    Capybara::Selenium::Driver.new(app, browser: :chrome, desired_capabilities: capabilities, http_client: client)
+    Capybara::Selenium::Driver.new(
```
[the remark this answers is not in the document]
Comment (srbarrios) [reply testsuite/features/support/env.rb:100]: This is only an indentation change to please Rubocop at the time I was working on this PR.

File: testsuite/features/support/env.rb (hunk 9 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb:153]: That's a leftover of the bigger refactor I made. This was necessary to recover properly when we run tests using parallel_tests gem, and some scenario fails.

I can revert it, and keep in another branch for now, still we gonna need it.

File: testsuite/features/support/env.rb (hunk 8 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb:204]: That's a leftover of the bigger refactor I made, to fully re-enable parallel tests. This was necessary to recover properly when we run tests using parallel_tests gem, and some scenario fails.

I can revert it, and keep in another branch for now, still we gonna need it.

File: testsuite/features/support/env.rb (hunk 10 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb:106]: It's part of the rest of refactor I was adding to facilitate running tests using parallel_tests gem, but I can revert that part and move it to another branch.

File: testsuite/features/support/env.rb (hunk 7 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb:224]: That gonna come later. For now this refactor will indeed cause the creation of a user in all the features, but that should not be a big deal.

The next PR I have in mind, @uyuni-project/qe for everyone awareness, is to organize the secondary features in a way that we can identify which ones can run in parallel and which ones don't support it. As for now, we can only identify it through our YAML files.

I'm open to proposals, my two PoCs gonna be : 
- A cucumber tag on the top of every feature: "@parallel" or "@sequential"
- Split secondary in two subfolders: "sequential" and "parallel"

File: testsuite/features/support/env.rb (hunk 9 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb]: Move into qe-refactor-enabling-parallel-tests branch

File: testsuite/features/support/env.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (srbarrios) [reply testsuite/features/support/env.rb]: Moved to qe-refactor-enabling-parallel-tests branch

File: testsuite/features/support/env.rb (hunk 10 above)
Comment (srbarrios) [reply testsuite/features/support/env.rb:106]: Moved to qe-refactor-enabling-parallel-tests branch

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 12 above)
Comment (srbarrios) [reply testsuite/features/step_definitions/navigation_steps.rb:381]: Using a different user, but keeping the same session it gonna require that we logout before login.

**Comment by srbarrios** [comment]
> Overall don't forget to update the copyright.

Good point.

**Review by maximenoel8 — APPROVED** [review]
